### PR TITLE
Error on download not found

### PR DESCRIPTION
**Pull request by nhynes**
This PR makes `download_*` return an error when the download is 404. Inexplicably, GCP does not return a JSON response, so it's difficult to convert the 404 into a `GoogleErrorResponse`. That's okay, though, because it's likely that other users of this library have their own version of

```rust
fn is_not_found_error(e: &cloud_storage::Error) -> bool {
    match e {
        cloud_storage::Error::Reqwest(e) if e.status() == Some(http::StatusCode::NOT_FOUND) => {
            true
        }
        cloud_storage::Error::Google(e) => {
            e.errors_has_reason(&cloud_storage::Reason::NotFound)
        }
        cloud_storage::Error::Other(msg) => msg.contains("No such object"),
        _ => false,
    }
}
```

Also, it'd be nice if this repo had codecov to allow directing effort towards additional tests.

### CODE REVIEW

File: src/resources/object.rs [hunk 1]
```diff
@@ -398,7 +398,11 @@ impl Object {
                 .send()
                 .await;
             let response = match response {
-                Ok(r) => r,
+                Ok(r) if r.status() == 200 => r,
+                Ok(r) => {
+                    let e = Error::Google(r.json().await.ok()?);
```
Comment (ThouCheese) [inline src/resources/object.rs:403]: Should we return `None` here when the JSON parsing fails? Wouldn't it make more sense to return an error?

File: src/resources/object.rs [hunk 2]
```diff
@@ -490,6 +494,8 @@ impl Object {
             .headers(crate::get_headers().await?)
             .send()
             .await?
+            .error_for_status()
+            .map_err(Error::from)?
```
Comment (ThouCheese) [inline src/resources/object.rs:498]: This explicit conversion is not needed, the ?-operator will do the conversion automatically.

File: src/resources/object.rs [hunk 3]
```diff
@@ -535,13 +541,15 @@ impl Object {
...
+        let response = crate::CLIENT
             .get(&url)
             .headers(crate::get_headers().await?)
             .send()
-            .await?;
-        let size = res.content_length();
-        let bytes = res
+            .await?
+            .error_for_status()
+            .map_err(Error::from)?;
```
Comment (ThouCheese) [inline src/resources/object.rs:550]: The same applies here.

**Review by ThouCheese — CHANGES_REQUESTED** [review]

**Comment by ThouCheese** [comment]
Hey @nhynes, sorry for only getting to this PR now. The premise of the PR makes a lot of sense, but I have a small question about how to best handle the JSON parsing error.

File: src/resources/object.rs (hunk 1 above)
Comment (nhynes) [reply src/resources/object.rs]: good idea. I think the entire batch of comments can be summarized as me not looking at the useful `From` implementations 😅

**Comment by ThouCheese** [comment]
LGTM

**Pull request merged**